Implement new status box design; add component previews

mnacos opened the pull request:
### Context

When viewing an application in the provider interface, it is important to highlight information about this application according to the current status of the application. New designs have been produced for summarising this information. The aim of this pull request is to implement the new status box designs.

### Changes proposed in this pull request

Modify the existing `provider_interface/status_box_component` to load subcomponents from within `provider_interface/status_box_components/` according to the current status of the application.

For example, the status box component will load `ProviderInterface::OfferComponent` to display the relevant status box.

Status boxes now take the entire available width of the column they are in, and general information about the application follows.

Status boxes also contain buttons and links for provider user actions, if these are allowed for the current status of an application.

There is also new code implementing `ActionView::Component` previews at `/rails/components`, also accessible through the Support Interface footer. The code, under `app/components/previews/`, will attempt to find applications matching the requested state within the current environments database, and complain if no appropriate data is found, rather than create mock data itself.

Component previews are only enabled on the same environments as mailer previews, i.e. development, qa and review apps.

### Guidance to review

Use http://localhost:3000/rails/components to preview the status box component for applications in different states.

The new status box takes up the whole width of the column at the top of the page, when viewing an application within the provider interface.

![image](https://user-images.githubusercontent.com/107591/74738501-f33cdb80-524e-11ea-9584-fa4edd86a4af.png)

This work also contains a `ConditionsComponent`, displaying a list of conditions with associated 'Pending, 'Met', 'Not met' labels, in preparation for marking individual conditions as Met/Not met.

![image](https://user-images.githubusercontent.com/107591/74738688-60e90780-524f-11ea-9704-8b57ded386a3.png)

### Link to Trello card

[Implement the new design for offers on the Provider UI application page](https://trello.com/c/dxIYA2hf)

## Things to check

- [x] This code doesn't rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training#azure-hosting-devops-pipeline)

duncanjbrown commented on app/components/provider_interface/course_presentation_component.rb:
```diff
@@ -0,0 +1,17 @@
+module ProviderInterface
+  class CoursePresentationComponent < ActionView::Component::Base
```
This (and LocationPresenterComponent) needs to be aware of the possibility that a different course could be offered, eg looking at `ApplicationChoice#offered_course`

mnacos replied:
Good catch, will make these components aware of this.

tijmenb commented:
Component previews also on https://apply-for-te-1635-imple-g8dt6g.herokuapp.com/rails/components

tijmenb commented on app/components/previews/provider_interface/status_box_component_preview.rb:
```diff
@@ -0,0 +1,59 @@
+module ProviderInterface
+  class StatusBoxComponentPreview < ActionView::Component::Preview
+    def awaiting_provider_decision
+      render_component_for ApplicationChoice.find_by(status: :awaiting_provider_decision)
```
https://apply-for-te-1635-imple-g8dt6g.herokuapp.com/rails/components/provider_interface/status_box_component/awaiting_provider_decision gives a 500 for me.

tijmenb commented on app/components/previews/provider_interface/status_box_component_preview.rb:
```diff
@@ -0,0 +1,59 @@
+      render_component_for ApplicationChoice.find_by(status: :awaiting_provider_decision)
+    end
+
+    def offer
```
https://apply-for-te-1635-imple-g8dt6g.herokuapp.com/rails/components/provider_interface/status_box_component/offer gives me a 500 too.

stevehook replied:
seems to work now

tijmenb commented on app/components/previews/provider_interface/status_box_component_preview.rb:
```diff
@@ -0,0 +1,59 @@
+    end
+
+    def rejected
+      render_component_for ApplicationChoice.find_by(status: :rejected)
```
Perhaps this to pick a different one each time?

```suggestion
      render_component_for ApplicationChoice.rejected.order('RANDOM()').first
```

duncanjbrown replied:
This preview renders a `withdrawn_by_us` application on my local setup — should we specify `offer_withdrawn_at` is `nil`?

duncanjbrown commented:
Component outcomes on Heroku after generating test applications:

application_withdrawn: 200
awaiting_provider_decision: 500
conditions_not_met: "Test data missing" error
declined: 200
enrolled: 200
offer: 500
offer_withdrawn: "Test data missing"
pending_conditions_with_conditions: 200
pending_conditions_with_no_conditions: "Test data missing"
recruited: 200
rejected: 200

tijmenb commented on app/components/provider_interface/conditions_component.html.erb:
```diff
@@ -1 +1,23 @@
-<%= render SummaryListComponent.new(rows: condition_rows) %>
+<% if condition_rows %>
```
Minor: you can get rid of this conditional by implementing `render?` in the component:

https://github.com/github/actionview-component/#conditional-rendering

mnacos replied:
Cool. I'll remove the defensive `&.` calls but implement `#render?` checking the application_choice is in the appropriate state.

duncanjbrown replied:
I think it should bail if the choice is in the wrong state as this means we have a bug in our calling code. It would be better to fail loudly than render nothing and us not know it was broken

tijmenb commented on app/components/provider_interface/status_box_components/recruited_component.html.erb:
```diff
@@ -0,0 +1,7 @@
+<%= render SummaryListComponent.new(rows: rows) %>
+<%= render ProviderInterface::ConditionsComponent.new(application_choice: application_choice) %>
+
+<a href="/application/2/confirm-enrollment" role="button" draggable="false" class="govuk-button" data-module="govuk-button">
```
This presumably needs to link to a real URL :)

mnacos replied:
:-) We don't let providers enrol candidates yet, removing...

tvararu commented:
This is really cool!

duncanjbrown commented on app/components/provider_interface/location_presentation_component.rb:
```diff
@@ -0,0 +1,22 @@
+
+    def site_address_or_name
+      address = site.full_address
+      address.presence || site.name
```
Is it odd not to show the site name, only the address, in this case?

mnacos replied:
Doing name + address, or just name, if address is missing

duncanjbrown commented on app/views/provider_interface/application_choices/show.html.erb:
```diff
@@ -19,7 +19,13 @@
+</div>
+
+<div class="govuk-grid-row">
+  <div class="govuk-grid-column">
```
I know there's nothing for the other third now, but I miss the `two-thirds` a bit. The lines feel very long

<img width="1044" alt="Screenshot 2020-02-18 at 17 13 40" src="https://user-images.githubusercontent.com/642279/74760260-090fc800-5272-11ea-8ac6-70372dca3874.png">

mnacos replied:
Bringing two-thirds back

duncanjbrown commented on app/views/layouts/_header.html.erb:
```diff
@@ -20,8 +20,10 @@
-                                 service_name: service_name,
-                                 service_url: service_link,
+                                 service_name: components_name || try(:service_name),
+                                 service_url: components_url || try(:service_url),
```
I wonder if there's a way to do this without it changing our production code. Could we somehow have a separate layout without worrying about getting out of sync with the proper frontend?

mnacos replied:
I'd rather not to introduce a separate layout... Current modifications are not too bad, I think, anything more and we'll have to add a separate layout. What do you think?

duncanjbrown replied:
Let's live with it in here rather than hold back this PR

mnacos commented on app/components/provider_interface/course_presentation_component.rb:
```diff
@@ -0,0 +1,25 @@
+    end
+
+    def course_option
+      application_choice.offered_course_option || application_choice.course_option
```
I think `OfferedCourseComponent` reduces the need for this.

mnacos commented on app/components/provider_interface/conditions_component.rb:
```diff
@@ -9,15 +9,23 @@ def initialize(application_choice:)
 
     def conditions
-      @application_choice.offer['conditions']
+      application_choice.offer['conditions'] if application_choice.offer.present?
```
I'll implement #render? on these components, which will take care of this.

duncanjbrown replied:
`#render?` looks good.

I think we can lose the `if` here as we never try to render this this for an application without an offer, and it's an error if we do

fofr commented:
@adamsilver Should the box be overhanging as it is in https://user-images.githubusercontent.com/107591/74738501-f33cdb80-524e-11ea-9584-fa4edd86a4af.png?

fofr commented on app/components/provider_interface/status_box_components/declined_component.html.erb:
```diff
@@ -0,0 +1,3 @@
+<h2 class="govuk-heading-l">Candidate declined the offer</h2>
```
I suggest: `[Name] declined your offer`

laurahermionetennant92 replied:
Yep that's better @fofr :)

fofr commented on app/components/provider_interface/status_box_components/enrolled_component.html.erb:
```diff
@@ -0,0 +1,3 @@
+<h2 class="govuk-heading-l">Candidate has enrolled on the course</h2>
```
Suggestion: `[Name] is enrolled`

fofr commented:
Can we update all titles that refer to "Candidate" with their name, the abstract is less useful.

fofr commented on app/components/provider_interface/status_box_components/pending_conditions_component.html.erb:
```diff
@@ -0,0 +1,7 @@
+<h2 class="govuk-heading-l">Candidate has accepted the offer</h2>
```
Suggestion: `[Name] accepted your offer`

fofr commented on app/components/provider_interface/status_box_components/recruited_component.html.erb:
```diff
@@ -0,0 +1,3 @@
+<h2 class="govuk-heading-l">Conditions of offer have been met</h2>
```
Suggestion: `[Name] has met the conditions of your offer`

adamsilver commented on app/components/provider_interface/status_box_components/conditions_not_met_component.rb:
```diff
@@ -0,0 +1,33 @@
+            value: render(ProviderInterface::ApplicationStatusTagComponent.new(application_choice: application_choice)),
+          },
+          {
+            key: 'Course',
```
We are looking to change this shortly so that provider name and course name are separate rows as they can be changed individually. This will more closely align with the work we're doing when the provider responds to the application initially - we're going to show radio buttons for make an offer, make an offer but change course, make an offer but change provider, make an offer but change location, reject application.

duncanjbrown replied:
@adamsilver I agree; I think they can still be separate rows even if we wrap them in a component though

adamsilver commented:
> https://user-images.githubusercontent.com/107591/74738501-f33cdb80-524e-11ea-9584-fa4edd86a4af.png

No it shouldn't.

mnacos commented:
All previews should work now on Heroku: https://apply-for-te-1635-imple-g8dt6g.herokuapp.com/rails/components

duncanjbrown commented:
> All previews should work now on Heroku: https://apply-for-te-1635-imple-g8dt6g.herokuapp.com/rails/components

Yep, looking good!

duncanjbrown commented on app/components/provider_interface/offered_course_component.html.erb:
```diff
@@ -0,0 +1,6 @@
+<% if display == :course %>
```
Having the logic in one component is great, I think we could go further and iron out the `display` param. It feels natural to have `OfferedCourseComponent` compose the rows for `Course` and `Location` rather than depend on callers to tell it which row to show?

duncanjbrown commented on app/components/provider_interface/status_box_component.rb:
```diff
@@ -8,54 +8,15 @@ def initialize(application_choice:)
-      date&.to_s(:govuk_date) || ''
-    end
+    # Should never happpen, but who knows
+    class ComponentMismatchError < StandardError; end
```
:+1:

duncanjbrown commented on app/workers/generate_test_applications.rb:
```diff
@@ -8,13 +8,16 @@ def perform
       TestApplications.create_application states: [:application_complete]
-      TestApplications.create_application states: [:awaiting_provider_decision]
-      TestApplications.create_application states: %i[offer offer]
+      TestApplications.create_application states: 3.times.map { :awaiting_provider_decision }
```
Or `[:awaiting_provider_decision] * 3` if you prefer?

duncanjbrown commented on spec/factories.rb:
```diff
@@ -218,6 +218,7 @@
       decline_by_default_at { Time.zone.now + 10.days }
       decline_by_default_days { 10 }
       offer { { 'conditions' => ['Be cool'] } }
+      offered_at { Time.zone.now }
```
Good spot!

mnacos commented:
> So close! The previews are brilliant. Two last questions:
> 
> * can we be even less tolerant in `ConditionsComponent`?
> * `OfferedCourseComponent` is nearly there but I think it wants one more refactor so it composes `Course` and `Location` rather than replacing them. Sorry - my bad, I should have written a better comment

No worries, I've updated the `ConditionsComponent` and removed `OfferedCourseComponent` entirely. As @adamsilver said, we'll need to split Provider and Course into separate rows, so I've just gone ahead and done that.

stevehook commented on app/components/provider_interface/status_box_components/awaiting_provider_decision_component.html.erb:
```diff
@@ -0,0 +1,3 @@
+<%= render SummaryListComponent.new(rows: rows) %>
```
Why doesn't this one have a title like all the others? It looked out of step with the others when I went through them on the review app - https://apply-for-te-1635-imple-g8dt6g.herokuapp.com/rails/components/provider_interface/status_box_component/awaiting_provider_decision

mnacos replied:
This is intentional / the current design. I'm guessing this is the first status that appears within the provider UI, anything past that point corresponds to a provider or candidate action we would like to highlight.
